PrefectHQ/prefect
# Update logging `JsonFormatter` to output valid JSON

This PR follows up on #7377 and updates `JsonFormatter` to output valid JSON. `JsonFormatter`'s current output is valid JavaScript, but not valid JSON:

```
{'name': 'prefect.flow_runs', 'msg': "Finished in state Completed('All states completed.')", 'args': [], 'levelname': 'INFO', 'levelno': 20, 'pathname': '/Users/adam/Library/Caches/pypoetry/virtualenvs/sett-prefect-vThVQ5x2-py3.9/lib/python3.9/site-packages/prefect/engine.py', 'filename': 'engine.py', 'module': 'engine', 'exc_info': None, 'exc_text': None, 'stack_info': None, 'lineno': 366, 'funcName': 'begin_flow_run', 'created': 1667049056.934597, 'msecs': 934.5970153808594, 'relativeCreated': 4141.36004447937, 'thread': 4377200000, 'threadName': 'MainThread', 'processName': 'MainProcess', 'process': 99513, 'flow_run_name': 'magenta-porpoise', 'flow_run_id': 'ee770a88-9a37-4ce3-b97b-5a3240cc6fab', 'flow_name': 'healthcheck', 'send_to_orion': False}
```

Consequently, *some* JSON parsers accept it but many (including DataDog's parser) do not. 

This PR updates the `JsonFormatter` to use Prefect's JSONSerializer and `orjson` to quickly serialize log entries to valid JSON like this:

```
{"name":"prefect.engine","msg":"Created flow run 'marvellous-aardwark' for flow 'my-flow'","args":[],"levelname":"INFO","levelno":20,"pathname":"/Users/ryanpeden/projects/prefect/prefect-azure-dev/prefect/src/prefect/engine.py","filename":"engine.py","module":"engine","exc_info":null,"exc_text":null,"stack_info":null,"lineno":222,"funcName":"create_then_begin_flow_run","created":1668703045.6771798,"msecs":677.1798133850098,"relativeCreated":1817.1977996826172,"thread":4373988736,"threadName":"MainThread","processName":"MainProcess","process":96887,"message":"Created flow run 'marvellous-aardwark' for flow 'my-flow'","asctime":"11:37:25"}
```

Closes #7377. 

### Example/Steps to QA

- Create a custom `logging.yml` based on the [default logging config YAML](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/logging/logging.yml) in ~/.prefect
- Note that `JsonFormatter` is already present in the default so you don't need to add it. 
- Add a handler in the `handlers` section of your `logging.yml`. For example:
- 
```
jsonfile:
        level: 0
        class: logging.handlers.RotatingFileHandler
        filename: /var/log/prefect.log
        # maximum 10MB per log file
        maxBytes: 10485760
        # keep the last 5 log files in addition to the current log file
        backupCount: 5
        formatter: json
```
* Add the new handler to one of the `loggers` defined in `logging.yml`. For example:
```
prefect.flow_runs:
        level: NOTSET
        handlers: [orion, console_flow_runs, jsonfile]
        propagate: no
```
* Run a Prefect flow
* Open the log filename defined earlier and observe that it contains correctly-formatted JSON log entries.


### Checklist


	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
